### PR TITLE
report: final metrics display, icons, whitespace polish

### PR DESCRIPTION
**Pull request by paulirish**
### screenshots
image
image

### example reports:

https://disastrous-bears.surge.sh/cnn.html
https://disastrous-bears.surge.sh/dbwtester.html
https://disastrous-bears.surge.sh/errors.html
https://disastrous-bears.surge.sh/example.html
https://disastrous-bears.surge.sh/paulirish.html
https://disastrous-bears.surge.sh/tinyhouse.html

https://disastrous-bears.surge.sh/z.devtools.cnn.html
https://disastrous-bears.surge.sh/z.devtools.dbwtester.html
https://disastrous-bears.surge.sh/z.devtools.errors.html
https://disastrous-bears.surge.sh/z.devtools.example.html
https://disastrous-bears.surge.sh/z.devtools.paulirish.html
https://disastrous-bears.surge.sh/z.devtools.tinyhouse.html

### CODE REVIEW

**Comment by vinamratasingal** [comment]
Looking good, LGTM on the UI side. 

QQ: is it normal for the tooltip to lag a bit? i.e. I hover for a few seconds, and then the tooltip shows up?

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 1]
```diff
@@ -120,12 +94,15 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
...
     this.createPermalinkSpan(element, category.id);
-    element.appendChild(this.renderCategoryScore(category));
+    element.appendChild(this.renderCategoryHeader(category));
 
+    // Metrics
     const metricAudits = category.audits.filter(audit => audit.group === 'metrics');
     const metricAuditsEl = this.renderAuditGroup(groups['metrics'], {expandable: false});
+    const estValuesEl = this.dom.createChildOf(metricAuditsEl, 'div',
+    'lh-metrics__disclaimer lh-metrics__disclaimer--head');
+    estValuesEl.textContent = 'Values are estimated and may vary. ';
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js]: aw I liked `Estimated values may vary`, I guess it doesn't necessarily convey that all the values you see are estimated though 🤔 

is the space at the end necessary? :)

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 2]
```diff
@@ -120,12 +94,15 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
...
     const element = this.dom.createElement('div', 'lh-category');
     this.createPermalinkSpan(element, category.id);
-    element.appendChild(this.renderCategoryScore(category));
+    element.appendChild(this.renderCategoryHeader(category));
 
+    // Metrics
     const metricAudits = category.audits.filter(audit => audit.group === 'metrics');
     const metricAuditsEl = this.renderAuditGroup(groups['metrics'], {expandable: false});
+    const estValuesEl = this.dom.createChildOf(metricAuditsEl, 'div',
+    'lh-metrics__disclaimer lh-metrics__disclaimer--head');
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js]: nit: indent

**Comment by paulirish** [comment]
> conflict party ahead 😭

all sorted now.

**Comment by paulirish** [comment]
@vinamratasingal 

> QQ: is it normal for the tooltip to lag a bit? i.e. I hover for a few seconds, and then the tooltip shows up?

I'm set them to be only show up after your cursor has been hovering there for 900ms. When the number is too low, you get unintentional tooltips.

I think the animation should take a little longer though. It's currently 150ms which is too fast given the long delay. 

Of course, all these parameters are tweakable. I do agree the tooltips feel a little funny atm. Hopefully we can tweak them to success.

**Comment by vinamratasingal** [comment]
@paulirish thanks for the explanation! SGTM on tweaking :)

File: lighthouse-core/report/html/renderer/category-renderer.js [hunk 3]
```diff
@@ -64,7 +64,7 @@ class CategoryRenderer {
       const tooltip = this.dom.createChildOf(textEl, 'div', 'lh-error-tooltip-content tooltip');
       tooltip.textContent = audit.result.debugString || 'Report error: no audit information';
     } else if (audit.result.debugString) {
-      const debugStrEl = auditEl.appendChild(this.dom.createElement('div', 'lh-debug'));
+      const debugStrEl = this.dom.createChildOf(titleEl, 'div', 'lh-debug');
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/category-renderer.js]: drive by restructuring? seems unrelated

File: lighthouse-core/report/html/renderer/category-renderer.js [hunk 4]
```diff
@@ -189,7 +190,7 @@ class CategoryRenderer {
     const notApplicableElem = this.renderAuditGroup({
       title: `Not applicable`,
     }, {expandable: true, itemCount: this._getTotalAuditsLength(elements)});
-    notApplicableElem.classList.add('lh-audit-group--notapplicable');
+    notApplicableElem.classList.add('lh-audit-group--adorned', 'lh-audit-group--notapplicable');
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/category-renderer.js]: a11y audits are the only unadorned audit-groups now then?

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 5]
```diff
@@ -88,18 +87,30 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
...
 
+  /**
+   * @param {!ReportRenderer.AuditJSON} audit
+   * @return {number|undefined}
+   */
+  _getWastedMs(audit) {
+    try {
+      return audit.result.details.summary.wastedMs;
+    } catch (e) {
+      return -1;
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js]: should this return undefined? it doesn't seem like the undefined case is really handled well now though, this is only for errors anyhoo right? maybe it always returns `wastedMs` or `MIN_VALUE` to force it to bottom or something?

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 6]
```diff
@@ -159,9 +177,11 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
       const groupEl = this.renderAuditGroup(groups['diagnostics'], {expandable: false});
       diagnosticAudits.forEach((item, i) => groupEl.appendChild(this.renderAudit(item, i)));
       groupEl.open = true;
+      groupEl.classList.add('lh-audit-group--adorned', 'lh-audit-group--diagnostics');
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js]: seems like the unadorned case is far more common, WDYT about it being the default to avoid this churn?

File: lighthouse-core/report/html/templates.html [hunk 7]
```diff
@@ -425,7 +425,7 @@ <h1 class="leftnav__header__title">Lighthouse</h1>
 <template id="tmpl-lh-gauge">
   <style>
     .lh-gauge__wrapper {
-      --circle-size: calc(2.5 * var(--header-font-size));
+      --circle-size: calc(3 * var(--header-font-size));
```
Comment (patrickhulce) [inline lighthouse-core/report/html/templates.html]: IDK Paul I'm feelin' a multiple more like  `2.7350215` ;)

File: lighthouse-core/report/html/renderer/category-renderer.js (hunk 3 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/category-renderer.js]: Yeah. Kinda separate but hey. I mentioned it on chat yesterday

before:
image

after:

image

File: lighthouse-core/report/html/renderer/category-renderer.js (hunk 4 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/category-renderer.js]: a11y and seo audit groups don't get icons, yeah.

File: lighthouse-core/report/html/renderer/performance-category-renderer.js (hunk 6 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: Not sure but i'll give it a shot.

File: lighthouse-core/report/html/renderer/category-renderer.js [hunk 8]
```diff
@@ -89,7 +89,7 @@ class CategoryRenderer {
    * @param {!ReportRenderer.CategoryJSON} category
    * @return {!Element}
    */
-  renderCategoryScore(category) {
+  renderCategoryHeader(category) {
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/category-renderer.js]: 👍 👍

File: lighthouse-core/test/report/html/renderer/category-renderer-test.js [hunk 9]
```diff
@@ -113,22 +113,6 @@ describe('CategoryRenderer', () => {
     category.description = prevDesc;
   });
 
-  it('renders audits with debugString as failed', () => {
```
Comment (brendankenny) [inline lighthouse-core/test/report/html/renderer/category-renderer-test.js]: do we have a test for this that handles rendering of the new modes? e.g. I don't see an error one

Comment (paulirish) [reply lighthouse-core/test/report/html/renderer/category-renderer-test.js]: tbh i'm not sure. :)

i'd like to do a followup once we have the 3 PRs landed (this one, scoreDisplayMode, and debugString -> warnings/errors/explanations) where we can try all the possibilities.

Comment (brendankenny) [reply lighthouse-core/test/report/html/renderer/category-renderer-test.js]: > i'd like to do a followup 

add to #5008?

File: lighthouse-core/report/html/renderer/performance-category-renderer.js (hunk 5 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: Yeah this is all about oppos that have errored.

I don't return undefined because of:

https://github.com/GoogleChrome/lighthouse/blob/cc5434b8f4a83fe37d1a9cb7841829031f113dc1/lighthouse-core/report/html/renderer/performance-category-renderer.js#L153-L154
image

min_value wfm!

**Comment by paulirish** [comment]
sg. done.​

**Comment by paulirish** [comment]
@brendankenny @patrickhulce rebased. can you rereview since this is front of the rest?

**Comment by paulirish** [comment]
And FWIW #5151 #5137 #5136 are rebased against imaginary branches right now because the conflicts just got REAL and i'll have to do the squash merge technique for them. :)

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 10]
```diff
@@ -86,18 +85,30 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
     return element;
   }
 
+  /**
+   * @param {!ReportRenderer.AuditJSON} audit
+   * @return {number|undefined}
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/performance-category-renderer.js]: not `|undefined` anymore

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 11]
```diff
@@ -124,15 +142,14 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
...
     // Opportunities
     const opportunityAudits = category.audits
         .filter(audit => audit.group === 'load-opportunities' && !Util.showAsPassed(audit.result))
-        .sort((auditA, auditB) => auditB.result.rawValue - auditA.result.rawValue);
+        .sort((auditA, auditB) => this._getWastedMs(auditB) - this._getWastedMs(auditA));
+
     if (opportunityAudits.length) {
-      const maxWaste = Math.max(...opportunityAudits.map(audit => audit.result.rawValue));
+      const wastedMsValues = opportunityAudits.map(audit => this._getWastedMs(audit));
+      const maxWaste = Math.max(...wastedMsValues);
```
[the remark this answers is not in the document]
Comment (paulirish) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: yes. only 1 visible opportunity and it was an error.

but it works out:

image

well. i guess we'd eventually set `elem.style.width = `NaN%`

:)

File: lighthouse-core/report/html/renderer/performance-category-renderer.js [hunk 12]
```diff
@@ -86,18 +85,30 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
     return element;
   }
 
+  /**
+   * @param {!ReportRenderer.AuditJSON} audit
+   * @return {number|undefined}
+   */
+  _getWastedMs(audit) {
+    try {
+      return audit.result.details.summary.wastedMs;
```
[the remark this answers is not in the document]
Comment (paulirish) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: I can't think of one, but I'm okay adding a typeof check

like adding this on the line above?:

```js
if (typeof audit.result.details.summary.wastedMs === 'number') return Number.MIN_VALUE;
```

File: lighthouse-core/report/html/renderer/performance-category-renderer.js (hunk 10 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: sure sg

File: lighthouse-core/report/html/renderer/performance-category-renderer.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: > I can't think of one, but I'm okay adding a typeof check

oh, I'm just wondering how you end up in the catch block

[the remark this answers is not in the document]
Comment (paulirish) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: don't opportunities with an error not have a summary block?

[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: ah, cool, so only in those cases? I was worried there were other weird cases where we were inconsistent. Maybe add a note that this is for fallback for errors?

File: lighthouse-core/report/html/renderer/performance-category-renderer.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: my preference for "real fix" would be to explicitly handle error cases separately so we can assume they're all regular sane numbers

but for now, I have no strong preference other than clearly expressing a nonsense number in the error case, MIN_VALUE seemed more obviously nonsense then -1, if it creates too much other noise compared to -1 though, I'm fine with it

File: lighthouse-core/report/html/renderer/performance-category-renderer.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/report/html/renderer/performance-category-renderer.js]: ok, maybe a more explicit `if (audit.result.details && audit.result.details.summary) {return audit.result.details.summary.wastedMs;} else { return Number.MIN_VALUE;}` then?

**Comment by paulirish** [comment]
thanks everyone for the reviews. appreciate it!
